Parent comment container

a-w-m opened the pull request.

a-w-m committed:
Merge branch 'authentication' of https://github.com/chingu-voyages/v16-bears-03-BE into Parent-Comment-Container

a-w-m committed:
Create Chat Window Parent Component
-Create comment component
-Create CreateComments component
-Create ViewCOmments component

a-w-m committed:
Merge branch 'dev' into Parent-Comment-Container

a-w-m committed:
Update state and props

a-w-m committed:
Merge branch 'dev' into Parent-Comment-Container

a-w-m committed:
Style Chat Window

a-w-m committed:
Fix Comment Rendering to Screen and Layout
-create grid layout for app container
-make ChatWindow stateful
-pass down state and dispatch from ChatWindow
-trigger useEffect in ViewComment when new comment created
-format date and time for comments
-fix scroll on ViewComments
-fix app layout

a-w-m committed:
Merge branch 'dev' into Parent-Comment-Container

a-w-m committed:
Merge branch 'dev' into Parent-Comment-Container

a-w-m commented on client/src/components/main/chatwindow/CreateComment.js:
```diff
@@ -0,0 +1,87 @@
+
+  const handleSubmit = text => {
+    axios
+      .post('/api/comments', { user: '5e4deef29b70942e38591863', text: text }, config)
```
I need to take user info out

rusye commented on client/src/components/main/AppContainer.js:
```diff
@@ -1,15 +1,20 @@
 
 function AppContainer() {
   return (
     <Container>
       <Sidebar />
+      <ChatWindow />
     </Container>
```
I say we delete this component and move this into the `App.js` file

Igor333m replied:
This is irrelevant to our project. But, the whole idea of components is to be reusable. So if we decided to use the sidebar and chat window somewhere else together, then we will need to refactor to look like this. We should break the app into more reusable components and try to keep things separate. Try to put separate logic into separate containers.

Igor333m replied:
If we delete this container, we will have Login, Register, Sidebar and main chat window in the same component. This is not the best practice in my opinion.

Igor333m replied:
```
> Even login and register should be in a separate container.
> <App>
>     <login&register>
>         <login>
>         <Register>
>     </login&register>
>     <appContainer>
>         <Sidebar>
>         <ChatWindow>
>     <appContainer>
> </App>
```

rusye commented on client/src/components/main/chatwindow/ChatWindow.js:
```diff
@@ -0,0 +1,33 @@
+*/
+
+const ChatWindow = props => {
+  const [newComment, setNewComment] = useState(false);
```
What is `[newComment, setNewComment]` and what is it supposed to do?

a-w-m replied:
I use `[newComment, setNewComment]` to track whether a new comment has been submitted by passing them down as props, triggering a re-render. It's in the parent component because both children components need to access the value. `CreateComment` uses `setNewComment` to change state, which initiates a re-render in `ViewComments`. Within `useEffect`, `ViewComments` uses `setNewComment` to change state again to trigger a second-rerender since the get request has a delay. When I was only changing state once through `setAllComments` in `ViewComments`, the response did not contain the new comment. I used the boolean `newComment` to trigger a second re-render, essentially a second get request to account for the delay.

a-w-m replied:
I've removed `useState` in favor of `useReducer` to clean things up

rusye commented on client/src/components/main/chatwindow/CreateComment.js:
```diff
@@ -0,0 +1,87 @@
+  };
+
+  const handleEnter = event => {
+    if (event.keyCode === 13) {
```
What does this do?

a-w-m replied:
this checks whether the enter key was pressed by checking for the keyCode value of the enter key

rusye commented on client/src/components/main/chatwindow/CreateComment.js:
```diff
@@ -0,0 +1,87 @@
+    setComment('');
+  };
+
+  const handleEnter = event => {
```
What's the difference between `handleEnter` and `handleSubmit`?

a-w-m replied:
`handleEnter` is an event handler that calls `handleSubmit` when the enter key is pressed.

`handleSubmit` is the event handler that processes the submit event on the create comment form. The submit event is posting the comment to the db.

So, together, when enter is pressed within the textarea input, submit is triggered.

rusye replied:
Could we have enter directly trigger `onSubmit`?

rusye commented on client/src/components/main/chatwindow/ViewComments.js:
```diff
@@ -0,0 +1,91 @@
+      {isError ? (
+        <div>"Something Went Wrong"</div>
+      ) : (
+        allComments.map(comment => {
```
Add a `key={index}` in each comment

a-w-m replied:
Gotcha, forgot to do this! Setting `key = comment._id`, since React recommends not using index if you have another unique identifier available.

a-w-m committed:
Merge branch 'dev' into Parent-Comment-Container

a-w-m committed:
Fix State Management
-implement useReducer and useContext hooks to manage state
-centralize user credentials in ChatWindow to make testingf easier
-adjust Comment component style
